### ISSUE TITLE
Rif Identify -- WalletConnect

### ISSUE THREAD

**Issue by AmazingBeerBelly** [rsksmart/rwallet#455]
流程：
1. Dapp端显示二维码，钱包端扫描二维码，然后选择某个钱包进行登录。钱包端调用approveSession方法就是允许登录，rejectSession方法就是拒绝登录，killSession方法是断开连接

`
  initWalletConnect = async () => {
    const { navigation } = this.props;
    const { uri } = navigation.state.params;

    this.setState({ loading: true });

    try {
      const connector = new WalletConnect({ uri });

      if (!connector.connected) {
        await connector.createSession();
      }

      console.log('connector: ', connector);

      await this.setState({
        loading: false,
        connector,
      });

      this.subscribeToEvents();
    } catch (error) {
      this.setState({ loading: false });

      throw error;
    }
  }
`

2. 当Dapp和钱包端建立连接之后，钱包端需要设置监听事件（subscribeToEvents方法）

`
  subscribeToEvents = () => {
    console.log('ACTION', 'subscribeToEvents');
    const { connector } = this.state;
    console.log('connector: ', connector);

    if (connector) {
      connector.on('session_request', (error, payload) => {
        console.log('EVENT', 'session_request');

        if (error) {
          throw error;
        }

        const { peerMeta } = payload.params[0];
        console.log('payload: ', payload);
        this.setState({ peerMeta });
      });

      connector.on('session_update', (error) => {
        console.log('EVENT', 'session_update');

        if (error) {
          throw error;
        }
      });

      connector.on('call_request', async (error, payload) => {
        const { method, params } = payload;
        // tslint:disable-next-line
        console.log('EVENT', 'call_request', 'method', method);
        console.log('EVENT', 'call_request', 'params', params);

        if (error) {
          throw error;
        }

        this.setState({ payload });
      });

      connector.on('connect', (error, payload) => {
        console.log('EVENT', 'connect');

        if (error) {
          throw error;
        }

        this.setState({ connected: true });
      });

      connector.on('disconnect', (error, payload) => {
        console.log('EVENT', 'disconnect');

        if (error) {
          throw error;
        }
      });

      if (connector.connected) {
        this.setState({
          connected: true,
        });
      }

      this.setState({ connector });
    }
  };
`

3. Dapp端发送一些操作（personal_sign, eth_sendTransaction等）给连接桥（Bridge Server），钱包端监听到这些操作之后，进行相对应的处理（就是对payload的method进行分别处理，在handleCallRequest方法中），然后调用approveRequest方法把结果返回给Dapp。调用rejectRequest则表示拒绝这些操作。

`
  import { ethers } from 'ethers';
  import Rsk3 from '@rsksmart/rsk3';

  initNetwork = async (network) => {
    this.rskEndpoint = network === 'Mainnet' ? MAINNET.RSK_END_POINT : TESTNET.RSK_END_POINT;
    this.provider = new ethers.providers.JsonRpcProvider(this.rskEndpoint);
    const chainId = network === 'Mainnet' ? MAINNET.NETWORK_VERSION : TESTNET.NETWORK_VERSION;
    this.setState({ chainId });
  }

  signMessage = async (signWallet) => {
    const { payload: { params } } = this.state;
    const message = Rsk3.utils.hexToAscii(params[0]);
    const signature = await signWallet.signMessage(message);
    return signature;
  }

  signTransaction = async (signWallet) => {
    const { selectedWallet: { address }, payload: { params } } = this.state;
    const nonce = await this.provider.getTransactionCount(Rsk3.utils.toChecksumAddress(address), 'pending');
    const gasPrice = await this.provider.getGasPrice();
    const txData = {
      nonce,
      data: params[0].data,
      gas: params[0].gas || 600000,
      gasPrice: params[0].gasPrice || gasPrice,
      to: Rsk3.utils.toChecksumAddress(params[0].to),
      value: (params[0].value && ethers.utils.bigNumberify(params[0].value)) || '0x0',
    };
    const rawTransaction = await signWallet.sign(txData);

    return rawTransaction;
  }

  sendRawTransaction = async (rawTransaction) => {
    const res = await this.provider.sendTransaction(rawTransaction);
    return res.hash;
  }

  handleCallRequest = async () => {
    const { selectedWallet: { privateKey }, payload } = this.state;
    const { id, method, params } = payload;

    let result = null;
    const signWallet = new ethers.Wallet(privateKey, this.provider);
    switch (method) {
      case 'personal_sign': {
        result = await this.signMessage(signWallet);
        break;
      }

      case 'eth_sign': {
        result = await this.signMessage(signWallet);
        break;
      }

      case 'eth_signTypedData': {
        break;
      }

      case 'eth_sendTransaction': {
        const rawTransaction = await this.signTransaction(signWallet);
        result = await this.sendRawTransaction(rawTransaction);
        break;
      }

      case 'eth_signTransaction': {
        result = await this.signTransaction(signWallet);
        break;
      }

      case 'eth_sendRawTransaction': {
        const rawTransaction = params[0];
        result = await this.sendRawTransaction(rawTransaction);
        break;
      }

      default:
        break;
    }

    return { id, result };
  }

  approveRequest = async () => {
    const { connector, payload } = this.state;
    const { id } = payload;

    try {
      const result = await this.handleCallRequest();
      connector.approveRequest(result);
    } catch (error) {
      console.error(error);
      if (connector) {
        connector.rejectRequest({
          id,
          error: { message: 'Failed or Rejected Request' },
        });
      }
    }

    this.closeRequest();
    await this.setState({ connector });
  }
`

**Comment by chrisli30** [comment]
Nice

**Comment by ilanolkies** [comment]
LRGTM! One question one advise:
- question: is the user asked before signing a transaction/mesaage? A pop up or a notification?
- advise: `eth_signTypedData` specs here https://eips.ethereum.org/EIPS/eip-712

**Comment by AmazingBeerBelly** [comment]
Yes, the user will be asked before signing a transaction/message. Usually, the wallet will pop a popup window to ask the user to approve or reject it.
=> question: is the user asked before signing a transaction/mesaage? A pop up or a notification?

Ok, thanks. I will spend some time researching eth_signTypedData
=> advise: eth_signTypedData specs here https://eips.ethereum.org/EIPS/eip-712

**Comment by grinry** [comment]
Is this already added to the app? Tried QR scanner in the last TestFlight build - from the app side it allows to connect, it even shows that the dapp is connected and allows to disconnect. However dapp itself does not receive any event about connection and just hangs with QR code opened, waiting for the connection.

**Comment by chrisli30** [comment]
Hi @grinry what dapp did you test against with? Is it on Testnet or Mainnet?

**Comment by grinry** [comment]
Hey @chrisli30, I tried both networks, dapp address is https://trade.sovryn.app/
Tested with ios (testflight) and android (build from source) with no luck
